## Telecom Churn Case study

### Business Problem
In the telecom industry, customers are able to choose from multiple service providers and actively switch from one operator to another. In this highly competitive market, the telecommunications industry experiences an average of 15-25% annual churn rate. Given the fact that it costs 5-10 times more to acquire a new customer than to retain an existing one, customer retention has now become even more important than customer acquisition.

For many incumbent operators, ___retaining high profitable customers is the number one business goal___.

To reduce customer churn, __telecom companies need to predict which customers are at high risk of churn__.

In this project, we will analyse customer-level data of a leading telecom firm, build predictive models to identify customers at high risk of churn and identify the main indicators of churn.

 __Importing Required Libraries__

In [144]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings as w
w.filterwarnings('ignore')
pd.set_option('max_columns',500)
pd.set_option('max_rows',500)

In [145]:
data=pd.read_csv('telecom_churn_data.csv')
data.head()

mobile_number  circle_id  loc_og_t2o_mou  std_og_t2o_mou  loc_ic_t2o_mou  \
0     7000842753        109             0.0             0.0             0.0   
1     7001865778        109             0.0             0.0             0.0   
2     7001625959        109             0.0             0.0             0.0   
3     7001204172        109             0.0             0.0             0.0   
4     7000142493        109             0.0             0.0             0.0   

  last_date_of_month_6 last_date_of_month_7 last_date_of_month_8  \
0            6/30/2014            7/31/2014            8/31/2014   
1            6/30/2014            7/31/2014            8/31/2014   
2            6/30/2014            7/31/2014            8/31/2014   
3            6/30/2014            7/31/2014            8/31/2014   
4            6/30/2014            7/31/2014            8/31/2014   

  last_date_of_month_9   arpu_6   arpu_7   arpu_8   arpu_9  onnet_mou_6  \
0            9/30/2014  197.385  214.816  213.803   21.100          NaN   
1            9/30/2014   34.047  355.074  268.321   86.285        24.11   
2            9/30/2014  167.690  189.058  210.226  290.714        11.54   
3            9/30/2014  221.338  251.102  508.054  389.500        99.91   
4            9/30/2014  261.636  309.876  238.174  163.426        50.31   

   onnet_mou_7  onnet_mou_8  onnet_mou_9  offnet_mou_6  offnet_mou_7  \
0          NaN         0.00          NaN           NaN           NaN   
1        78.68         7.68        18.34         15.74         99.84   
2        55.24        37.26        74.81        143.33        220.59   
3        54.39       310.98       241.71        123.31        109.01   
4       149.44        83.89        58.78         76.96         91.88   

   offnet_mou_8  offnet_mou_9  roam_ic_mou_6  roam_ic_mou_7  roam_ic_mou_8  \
0          0.00           NaN            NaN            NaN           0.00   
1        304.76         53.76            0.0           0.00           0.00   
2        208.36        118.91            0.0           0.00           0.00   
3         71.68        113.54            0.0          54.86          44.38   
4        124.26         45.81            0.0           0.00           0.00   

   roam_ic_mou_9  roam_og_mou_6  roam_og_mou_7  roam_og_mou_8  roam_og_mou_9  \
0            NaN            NaN            NaN           0.00            NaN   
1           0.00            0.0           0.00           0.00           0.00   
2          38.49            0.0           0.00           0.00          70.94   
3           0.00            0.0          28.09          39.04           0.00   
4           0.00            0.0           0.00           0.00           0.00   

   loc_og_t2t_mou_6  loc_og_t2t_mou_7  loc_og_t2t_mou_8  loc_og_t2t_mou_9  \
0               NaN               NaN              0.00               NaN   
1             23.88             74.56              7.68             18.34   
2              7.19             28.74             13.58             14.39   
3             73.68             34.81             10.61             15.49   
4             50.31            149.44             83.89             58.78   

   loc_og_t2m_mou_6  loc_og_t2m_mou_7  loc_og_t2m_mou_8  loc_og_t2m_mou_9  \
0               NaN               NaN              0.00               NaN   
1             11.51             75.94            291.86             53.76   
2             29.34             16.86             38.46             28.16   
3            107.43             83.21             22.46             65.46   
4             67.64             91.88            124.26             37.89   

   loc_og_t2f_mou_6  loc_og_t2f_mou_7  loc_og_t2f_mou_8  loc_og_t2f_mou_9  \
0               NaN               NaN              0.00               NaN   
1              0.00              0.00              0.00              0.00   
2             24.11             21.79             15.61             22.24   
3              1.91              0.65              

In [146]:
data.shape

(99999, 226)

In [147]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99999 entries, 0 to 99998
Columns: 226 entries, mobile_number to sep_vbc_3g
dtypes: float64(179), int64(35), object(12)
memory usage: 172.4+ MB


As checked above, there are 214 numeric columns and 12 non-numeric columns

In [148]:
# look at data statistics
data.describe(include='all')

mobile_number  circle_id  loc_og_t2o_mou  std_og_t2o_mou  \
count    9.999900e+04    99999.0         98981.0         98981.0   
unique            NaN        NaN             NaN             NaN   
top               NaN        NaN             NaN             NaN   
freq              NaN        NaN             NaN             NaN   
mean     7.001207e+09      109.0             0.0             0.0   
std      6.956694e+05        0.0             0.0             0.0   
min      7.000000e+09      109.0             0.0             0.0   
25%      7.000606e+09      109.0             0.0             0.0   
50%      7.001205e+09      109.0             0.0             0.0   
75%      7.001812e+09      109.0             0.0             0.0   
max      7.002411e+09      109.0             0.0             0.0   

        loc_ic_t2o_mou last_date_of_month_6 last_date_of_month_7  \
count          98981.0                99999                99398   
unique             NaN                    1                    1   
top                NaN            6/30/2014            7/31/2014   
freq               NaN                99999                99398   
mean               0.0                  NaN                  NaN   
std                0.0                  NaN                  NaN   
min                0.0                  NaN                  NaN   
25%                0.0                  NaN                  NaN   
50%                0.0                  NaN                  NaN   
75%                0.0                  NaN                  NaN   
max                0.0                  NaN                  NaN   

       last_date_of_month_8 last_date_of_month_9        arpu_6        arpu_7  \
count                 98899                98340  99999.000000  99999.000000   
unique                    1                    1           NaN           NaN   
top               8/31/2014            9/30/2014           NaN           NaN   
freq                  98899                98340           NaN           NaN   
mean                    NaN                  NaN    282.987358    278.536648   
std                     NaN                  NaN    328.439770    338.156291   
min                     NaN                  NaN  -2258.709000  -2014.045000   
25%                     NaN                  NaN     93.411500     86.980500   
50%                     NaN                  NaN    197.704000    191.640000   
75%                     NaN                  NaN    371.060000    365.344500   
max                     NaN                  NaN  27731.088000  35145.834000   

              arpu_8        arpu_9   onnet_mou_6   onnet_mou_7   onnet_mou_8  \
count   99999.000000  99999.000000  96062.000000  96140.000000  94621.000000   
unique           NaN           NaN           NaN           NaN           NaN   
top              NaN           NaN           NaN           NaN           NaN   
freq             NaN           NaN           NaN           NaN           NaN   
mean      279.154731    261.645069    132.395875    133.670805    133.018098   
std       344.474791    341.998630    297.207406    308.794148    308.951589   
min      -945.808000  -1899.505000      0.000000      0.000000      0.000000   
25%        84.126000     62.685000      7.380000      6.660000      6.460000   
50%       192.080000    176.849000     34.310000     32.330000     32.360000   
75%       369.370500    353.466500    118.740000    115.595000    115.860000   
max     33543.624000  38805.617000   7376.710000   8157.780000  10752.560000   

         onnet_mou_9  offnet_mou_6  offnet_mou_7  offnet_mou_8  offnet_mou_9  \
count   92254.000000  96062.000000  96140.000000  94621.000000  92254.000000   
unique           NaN           NaN           NaN           NaN           NaN   
top              NaN           NaN           NaN           NaN           NaN   
freq             NaN           NaN           NaN           NaN           NaN   
mean      130.302327    197.935577    197.045133    

#### In churn prediction, we assume that there are three phases of customer lifecycle :

The ‘good’ phase [Month 6 & 7]<br>
The ‘action’ phase [Month 8]<br>
The ‘churn’ phase [Month 9]<br><br>
In this case, since we are working over a four-month window, the first two months are the ‘good’ phase, the third month is the ‘action’ phase, while the fourth month is the ‘churn’ phase.

### Data Preparation

Let us create some utility functions

In [149]:
# Method for Checking missing values percentages
def checkMissingPercent(dataset, cutoff):
    missing = round(100*(dataset.isnull().sum()/dataset.shape[0]))
    return missing.loc[missing>cutoff]

In [150]:
# Method for imputing data 
def imputeData(df, col_list):
    for i in [x + y for y in ['_6','_7','_8','_9'] for x in col_list]:
        df[i].fillna(0,inplace = True)

__Handling missing values__

In [151]:
mod_data=data.copy()

In [152]:
# Since mobile no has all unique values and represents a particular customer, it can be dropped from the dataset.
# Similarly, circle_id has all same values(109), it also can be dropped.
mod_data.drop(['mobile_number', 'circle_id'], axis=1, inplace=True)

In [153]:
# look at missing value ratio in each column
checkMissingPercent(mod_data, 0)

loc_og_t2o_mou               1.0
std_og_t2o_mou               1.0
loc_ic_t2o_mou               1.0
last_date_of_month_7         1.0
last_date_of_month_8         1.0
last_date_of_month_9         2.0
onnet_mou_6                  4.0
onnet_mou_7                  4.0
onnet_mou_8                  5.0
onnet_mou_9                  8.0
offnet_mou_6                 4.0
offnet_mou_7                 4.0
offnet_mou_8                 5.0
offnet_mou_9                 8.0
roam_ic_mou_6                4.0
roam_ic_mou_7                4.0
roam_ic_mou_8                5.0
roam_ic_mou_9                8.0
roam_og_mou_6                4.0
roam_og_mou_7                4.0
roam_og_mou_8                5.0
roam_og_mou_9                8.0
loc_og_t2t_mou_6             4.0
loc_og_t2t_mou_7             4.0
loc_og_t2t_mou_8             5.0
loc_og_t2t_mou_9             8.0
loc_og_t2m_mou_6             4.0
loc_og_t2m_mou_7             4.0
loc_og_t2m_mou_8             5.0
loc_og_t2m_mou_9             8.0
loc_og_t2f

As checked above, there are so many columns conatining missing values. Among them, there are some columns which has more than 70% of missing values. We will not directly delete those columns. Let us first check that these values as null because of no transactions or because of some other reason.

In [154]:
# getting all columns for month of June which has 75% missing values
cols = checkMissingPercent(mod_data, 74).index

mod_data.loc[mod_data.date_of_last_rech_data_6.isna(),cols].head()

date_of_last_rech_data_6  total_rech_data_6  max_rech_data_6  \
1                      NaN                NaN              NaN   
2                      NaN                NaN              NaN   
3                      NaN                NaN              NaN   
5                      NaN                NaN              NaN   
6                      NaN                NaN              NaN   

   count_rech_2g_6  count_rech_3g_6  av_rech_amt_data_6  arpu_3g_6  arpu_2g_6  \
1              NaN              NaN                 NaN        NaN        NaN   
2              NaN              NaN                 NaN        NaN        NaN   
3              NaN              NaN                 NaN        NaN        NaN   
5              NaN              NaN                 NaN        NaN        NaN   
6              NaN              NaN                 NaN        NaN        NaN   

   night_pck_user_6  fb_user_6  
1               NaN        NaN  
2               NaN        NaN  
3               NaN        NaN  
5               NaN        NaN  
6               NaN        NaN

As checked above, all the columns has null values where date of last recharge is missing. This is valid, we can replace these null values with 0 as there is no recharge done.

In [155]:
# imputing all the columns other than those containg date with 0 having more than 50% missing value
cols = list(filter(lambda x : not x.startswith('date') , checkMissingPercent(mod_data, 50).index))

mod_data[cols]=mod_data[cols].apply(lambda x: x.fillna(0))
mod_data[cols].head()

total_rech_data_6  total_rech_data_7  total_rech_data_8  total_rech_data_9  \
0                1.0                1.0                1.0                0.0   
1                0.0                1.0                2.0                0.0   
2                0.0                0.0                0.0                1.0   
3                0.0                0.0                0.0                0.0   
4                1.0                0.0                0.0                0.0   

   max_rech_data_6  max_rech_data_7  max_rech_data_8  max_rech_data_9  \
0            252.0            252.0            252.0              0.0   
1              0.0            154.0             25.0              0.0   
2              0.0              0.0              0.0             46.0   
3              0.0              0.0              0.0              0.0   
4             56.0              0.0              0.0              0.0   

   count_rech_2g_6  count_rech_2g_7  count_rech_2g_8  count_rech_2g_9  \
0              0.0              0.0              0.0              0.0   
1              0.0              1.0              2.0              0.0   
2              0.0              0.0              0.0              1.0   
3              0.0              0.0              0.0              0.0   
4              1.0              0.0              0.0              0.0   

   count_rech_3g_6  count_rech_3g_7  count_rech_3g_8  count_rech_3g_9  \
0              1.0              1.0              1.0              0.0   
1              0.0              0.0              0.0              0.0   
2              0.0              0.0              0.0              0.0   
3              0.0              0.0              0.0              0.0   
4              0.0              0.0              0.0              0.0   

   av_rech_amt_data_6  av_rech_amt_data_7  av_rech_amt_data_8  \
0               252.0               252.0               252.0   
1                 0.0               154.0                50.0   
2                 0.0                 0.0                 0.0   
3                 0.0                 0.0                 0.0   
4                56.0                 0.0                 0.0   

   av_rech_amt_data_9  arpu_3g_6  arpu_3g_7  arpu_3g_8  arpu_3g_9  arpu_2g_6  \
0                 0.0     212.17     212.17     212.17       0.00     212.17   
1                 0.0       0.00       0.00       0.00       0.00       0.00   
2                46.0       0.00       0.00       0.00       2.84       0.00   
3                 0.0       0.00       0.00       0.00       0.00       0.00   
4                 0.0       0.00       0.00       0.00       0.00       0.00   

   arpu_2g_7  arpu_2g_8  arpu_2g_9  night_pck_user_6  night_pck_user_7  \
0     212.17     212.17        0.0               0.0               0.0   
1      28.61       7.60        0.0               0.0               0.0   
2       0.00       0.00        0.0               0.0               0.0   
3       0.00       0.00        0.0               0.0               0.0   
4       0.00       0.00        0.0               0.0               0.0   

   night_pck_user_8  night_pck_user_9  fb_user_6  fb_user_7  fb_user_8  \
0               0.0               0.0        1.0        1.0        1.0   
1               0.0               0.0        0.0        1.0        1.0   
2               0.0               0.0        0.0        0.0        0.0   
3               0.0               0.0        0.0        0.0        0.0   
4               0.0               0.0        0.0        0.0        0.0   

   fb_user_9  
0        0.0  
1        0.0  
2        1.0  
3        0.0  
4        0.0

Checking again percent of missing values

In [156]:
checkMissingPercent(mod_data, 0)

loc_og_t2o_mou               1.0
std_og_t2o_mou               1.0
loc_ic_t2o_mou               1.0
last_date_of_month_7         1.0
last_date_of_month_8         1.0
last_date_of_month_9         2.0
onnet_mou_6                  4.0
onnet_mou_7                  4.0
onnet_mou_8                  5.0
onnet_mou_9                  8.0
offnet_mou_6                 4.0
offnet_mou_7                 4.0
offnet_mou_8                 5.0
offnet_mou_9                 8.0
roam_ic_mou_6                4.0
roam_ic_mou_7                4.0
roam_ic_mou_8                5.0
roam_ic_mou_9                8.0
roam_og_mou_6                4.0
roam_og_mou_7                4.0
roam_og_mou_8                5.0
roam_og_mou_9                8.0
loc_og_t2t_mou_6             4.0
loc_og_t2t_mou_7             4.0
loc_og_t2t_mou_8             5.0
loc_og_t2t_mou_9             8.0
loc_og_t2m_mou_6             4.0
loc_og_t2m_mou_7             4.0
loc_og_t2m_mou_8             5.0
loc_og_t2m_mou_9             8.0
loc_og_t2f

Let us have a look at the non-numeric columns

In [157]:
obj=mod_data.select_dtypes(include='object')
for i in obj.columns:
    print(i,'', obj[i].nunique(),'', obj[i].isna().sum()) 

last_date_of_month_6  1  0
last_date_of_month_7  1  601
last_date_of_month_8  1  1100
last_date_of_month_9  1  1659
date_of_last_rech_6  30  1607
date_of_last_rech_7  31  1767
date_of_last_rech_8  31  3622
date_of_last_rech_9  30  4760
date_of_last_rech_data_6  30  74846
date_of_last_rech_data_7  31  74428
date_of_last_rech_data_8  31  73660
date_of_last_rech_data_9  30  74077


We have already used date to fill the missing values. Further these date columns seems to be irrelevant in our analysis, so we will drop these columns

In [158]:
mod_data = mod_data.drop(obj.columns, axis=1)

Again checking for the missing values

In [159]:
checkMissingPercent(mod_data, 0)

loc_og_t2o_mou      1.0
std_og_t2o_mou      1.0
loc_ic_t2o_mou      1.0
onnet_mou_6         4.0
onnet_mou_7         4.0
onnet_mou_8         5.0
onnet_mou_9         8.0
offnet_mou_6        4.0
offnet_mou_7        4.0
offnet_mou_8        5.0
offnet_mou_9        8.0
roam_ic_mou_6       4.0
roam_ic_mou_7       4.0
roam_ic_mou_8       5.0
roam_ic_mou_9       8.0
roam_og_mou_6       4.0
roam_og_mou_7       4.0
roam_og_mou_8       5.0
roam_og_mou_9       8.0
loc_og_t2t_mou_6    4.0
loc_og_t2t_mou_7    4.0
loc_og_t2t_mou_8    5.0
loc_og_t2t_mou_9    8.0
loc_og_t2m_mou_6    4.0
loc_og_t2m_mou_7    4.0
loc_og_t2m_mou_8    5.0
loc_og_t2m_mou_9    8.0
loc_og_t2f_mou_6    4.0
loc_og_t2f_mou_7    4.0
loc_og_t2f_mou_8    5.0
loc_og_t2f_mou_9    8.0
loc_og_t2c_mou_6    4.0
loc_og_t2c_mou_7    4.0
loc_og_t2c_mou_8    5.0
loc_og_t2c_mou_9    8.0
loc_og_mou_6        4.0
loc_og_mou_7        4.0
loc_og_mou_8        5.0
loc_og_mou_9        8.0
std_og_t2t_mou_6    4.0
std_og_t2t_mou_7    4.0
std_og_t2t_mou_8

In [160]:
cols=list(checkMissingPercent(mod_data, 0).index)
mod_data[cols].describe()

loc_og_t2o_mou  std_og_t2o_mou  loc_ic_t2o_mou   onnet_mou_6  \
count         98981.0         98981.0         98981.0  96062.000000   
mean              0.0             0.0             0.0    132.395875   
std               0.0             0.0             0.0    297.207406   
min               0.0             0.0             0.0      0.000000   
25%               0.0             0.0             0.0      7.380000   
50%               0.0             0.0             0.0     34.310000   
75%               0.0             0.0             0.0    118.740000   
max               0.0             0.0             0.0   7376.710000   

        onnet_mou_7   onnet_mou_8   onnet_mou_9  offnet_mou_6  offnet_mou_7  \
count  96140.000000  94621.000000  92254.000000  96062.000000  96140.000000   
mean     133.670805    133.018098    130.302327    197.935577    197.045133   
std      308.794148    308.951589    308.477668    316.851613    325.862803   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        6.660000      6.460000      5.330000     34.730000     32.190000   
50%       32.330000     32.360000     29.840000     96.310000     91.735000   
75%      115.595000    115.860000    112.130000    231.860000    226.815000   
max     8157.780000  10752.560000  10427.460000   8362.360000   9667.130000   

       offnet_mou_8  offnet_mou_9  roam_ic_mou_6  roam_ic_mou_7  \
count  94621.000000  92254.000000   96062.000000   96140.000000   
mean     196.574803    190.337222       9.950013       7.149898   
std      327.170662    319.396092      72.825411      73.447948   
min        0.000000      0.000000       0.000000       0.000000   
25%       31.630000     27.130000       0.000000       0.000000   
50%       92.140000     87.290000       0.000000       0.000000   
75%      228.260000    220.505000       0.000000       0.000000   
max    14007.340000  10310.760000   13724.380000   15371.040000   

       roam_ic_mou_8  roam_ic_mou_9  roam_og_mou_6  roam_og_mou_7  \
count   94621.000000   92254.000000   96062.000000   96140.000000   
mean        7.292981       6.343841      13.911337       9.818732   
std        68.402466      57.137537      71.443196      58.455762   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       0.000000       0.000000   
75%         0.000000       0.000000       0.000000       0.000000   
max     13095.360000    8464.030000    3775.110000    2812.040000   

       roam_og_mou_8  roam_og_mou_9  loc_og_t2t_mou_6  loc_og_t2t_mou_7  \
count   94621.000000   92254.000000      96062.000000      96140.000000   
mean        9.971890       8.555519         47.100763         46.473010   
std        64.713221      58.438186        150.856393        155.318705   
min         0.000000       0.000000          0.000000          0.000000   
25%         0.000000       0.000000          1.660000          1.630000   
50%         0.000000       0.000000         11.910000         11.610000   
75%         0.000000       0.000000         40.960000         39.910000   
max      5337.040000    4428.460000       6431.330000       7400.660000   

       loc_og_t2t_mou_8  loc_og_t2t_mou_9  loc_og_t2m_mou_6  loc_og_t2m_mou_7  \
count      94621.000000      92254.000000      96062.000000      96140.000000   
mean          45.887806         44.584446         93.342088         91.397131   
std          151.184830        147.995390        162.780544        157.492308   
min            0.000000          0.000000          0.000000          0.000000   
25%            1.600000          1.360000          9.880000         10.025000   
50%           11.730000         11.260000         41.030000         40.430000   
75%           40.110000         39.280000        110.390000        107.560000   
max        10752.560000      10389.240000       4729.740000       4557.140000   

       loc_og_t2m_mou_8

As checked above, all the columns have their minimum value 0, but since the missing percent is very low around 4-5%, this can be because of technical or human error, its better to fill these values with median rather than 0. 

In [161]:
# filling the columns above with median
mod_data[cols]=mod_data[cols].apply(lambda x: x.fillna(x.median()))
mod_data[cols].head()

loc_og_t2o_mou  std_og_t2o_mou  loc_ic_t2o_mou  onnet_mou_6  onnet_mou_7  \
0             0.0             0.0             0.0        34.31        32.33   
1             0.0             0.0             0.0        24.11        78.68   
2             0.0             0.0             0.0        11.54        55.24   
3             0.0             0.0             0.0        99.91        54.39   
4             0.0             0.0             0.0        50.31       149.44   

   onnet_mou_8  onnet_mou_9  offnet_mou_6  offnet_mou_7  offnet_mou_8  \
0         0.00        29.84         96.31        91.735          0.00   
1         7.68        18.34         15.74        99.840        304.76   
2        37.26        74.81        143.33       220.590        208.36   
3       310.98       241.71        123.31       109.010         71.68   
4        83.89        58.78         76.96        91.880        124.26   

   offnet_mou_9  roam_ic_mou_6  roam_ic_mou_7  roam_ic_mou_8  roam_ic_mou_9  \
0         87.29            0.0           0.00           0.00           0.00   
1         53.76            0.0           0.00           0.00           0.00   
2        118.91            0.0           0.00           0.00          38.49   
3        113.54            0.0          54.86          44.38           0.00   
4         45.81            0.0           0.00           0.00           0.00   

   roam_og_mou_6  roam_og_mou_7  roam_og_mou_8  roam_og_mou_9  \
0            0.0           0.00           0.00           0.00   
1            0.0           0.00           0.00           0.00   
2            0.0           0.00           0.00          70.94   
3            0.0          28.09          39.04           0.00   
4            0.0           0.00           0.00           0.00   

   loc_og_t2t_mou_6  loc_og_t2t_mou_7  loc_og_t2t_mou_8  loc_og_t2t_mou_9  \
0             11.91             11.61              0.00             11.26   
1             23.88             74.56              7.68             18.34   
2              7.19             28.74             13.58             14.39   
3             73.68             34.81             10.61             15.49   
4             50.31            149.44             83.89             58.78   

   loc_og_t2m_mou_6  loc_og_t2m_mou_7  loc_og_t2m_mou_8  loc_og_t2m_mou_9  \
0             41.03             40.43              0.00             39.12   
1             11.51             75.94            291.86             53.76   
2             29.34             16.86             38.46             28.16   
3            107.43             83.21             22.46             65.46   
4             67.64             91.88            124.26             37.89   

   loc_og_t2f_mou_6  loc_og_t2f_mou_7  loc_og_t2f_mou_8  loc_og_t2f_mou_9  \
0              0.00              0.00              0.00              0.00   
1              0.00              0.00              0.00              0.00   
2             24.11             21.79             15.61             22.24   
3              1.91              0.65              4.91              2.06   
4              0.00              0.00              0.00              1.93   

   loc_og_t2c_mou_6  loc_og_t2c_mou_7  loc_og_t2c_mou_8  loc_og_t2c_mou_9  \
0               0.0              0.00              0.00              0.00   
1               0.0              2.91              0.00              0.00   
2               0.0            135.54             45.76              0.48   
3               0.0              0.00              0.00              0.00   
4               0.0              0.00              0.00              0.00   

   loc_og_mou_6  loc_og_mou_7  loc_og_mou_8  loc_og_mou_9  std_og_t2t_mou_6  \
0         65.11        63.685          0.00         61.84              0.00   
1         35.39       150.510        299.54         72.11              0.23   
2         60.66        67.410         67.66         64.81              4.34   
3        183.03       118.680         37.99         83.

Checking if our missing value imputation is successfully done or not

In [162]:
all(mod_data.isna().sum()==0)

True

In [163]:
# removing duplicates from row
mod_data.drop_duplicates(inplace=True)
mod_data.shape

(99999, 212)

There are a few columns whose names are not consistent with other columns. Let make them same.

In [164]:
print(list(filter(lambda x: x[-1].isalpha(), mod_data.columns)))
mod_data.rename(columns={'aug_vbc_3g':'vbc_3g_8', 'jul_vbc_3g':'vbc_3g_7', 'jun_vbc_3g':'vbc_3g_6'}, inplace=True)
mod_data.head()

['loc_og_t2o_mou', 'std_og_t2o_mou', 'loc_ic_t2o_mou', 'aon', 'aug_vbc_3g', 'jul_vbc_3g', 'jun_vbc_3g', 'sep_vbc_3g']


loc_og_t2o_mou  std_og_t2o_mou  loc_ic_t2o_mou   arpu_6   arpu_7   arpu_8  \
0             0.0             0.0             0.0  197.385  214.816  213.803   
1             0.0             0.0             0.0   34.047  355.074  268.321   
2             0.0             0.0             0.0  167.690  189.058  210.226   
3             0.0             0.0             0.0  221.338  251.102  508.054   
4             0.0             0.0             0.0  261.636  309.876  238.174   

    arpu_9  onnet_mou_6  onnet_mou_7  onnet_mou_8  onnet_mou_9  offnet_mou_6  \
0   21.100        34.31        32.33         0.00        29.84         96.31   
1   86.285        24.11        78.68         7.68        18.34         15.74   
2  290.714        11.54        55.24        37.26        74.81        143.33   
3  389.500        99.91        54.39       310.98       241.71        123.31   
4  163.426        50.31       149.44        83.89        58.78         76.96   

   offnet_mou_7  offnet_mou_8  offnet_mou_9  roam_ic_mou_6  roam_ic_mou_7  \
0        91.735          0.00         87.29            0.0           0.00   
1        99.840        304.76         53.76            0.0           0.00   
2       220.590        208.36        118.91            0.0           0.00   
3       109.010         71.68        113.54            0.0          54.86   
4        91.880        124.26         45.81            0.0           0.00   

   roam_ic_mou_8  roam_ic_mou_9  roam_og_mou_6  roam_og_mou_7  roam_og_mou_8  \
0           0.00           0.00            0.0           0.00           0.00   
1           0.00           0.00            0.0           0.00           0.00   
2           0.00          38.49            0.0           0.00           0.00   
3          44.38           0.00            0.0          28.09          39.04   
4           0.00           0.00            0.0           0.00           0.00   

   roam_og_mou_9  loc_og_t2t_mou_6  loc_og_t2t_mou_7  loc_og_t2t_mou_8  \
0           0.00             11.91             11.61              0.00   
1           0.00             23.88             74.56              7.68   
2          70.94              7.19             28.74             13.58   
3           0.00             73.68             34.81             10.61   
4           0.00             50.31            149.44             83.89   

   loc_og_t2t_mou_9  loc_og_t2m_mou_6  loc_og_t2m_mou_7  loc_og_t2m_mou_8  \
0             11.26             41.03             40.43              0.00   
1             18.34             11.51             75.94            291.86   
2             14.39             29.34             16.86             38.46   
3             15.49            107.43             83.21             22.46   
4             58.78             67.64             91.88            124.26   

   loc_og_t2m_mou_9  loc_og_t2f_mou_6  loc_og_t2f_mou_7  loc_og_t2f_mou_8  \
0             39.12              0.00              0.00              0.00   
1             53.76              0.00              0.00              0.00   
2             28.16             24.11             21.79             15.61   
3             65.46              1.91              0.65              4.91   
4             37.89              0.00              0.00              0.00   

   loc_og_t2f_mou_9  loc_og_t2c_mou_6  loc_og_t2c_mou_7  loc_og_t2c_mou_8  \
0              0.00               0.0              0.00              0.00   
1              0.00               0.0              2.91              0.00   
2             22.24               0.0            135.54             45.76   
3              2.06               0.0              0.00              0.00   
4              1.93               0.0              0.00              0.00   

   loc_og_t2c_mou_9  loc_og_mou_6  loc_og_mou_7  loc_og_mou_8  loc_og_mou_9  \
0              0.00         65.11        63.685          0.00         61.84   
1              0.00         35.39       150.510        299.54         72.11   
2              0.48       

__Taking only the data of high valued customer by taking average of total recharge amount of good months__

In [165]:
mod_data['av_rech_amt_6_7']=((mod_data.av_rech_amt_data_6 * mod_data.total_rech_data_6 + mod_data.total_rech_amt_6)+
                             (mod_data.av_rech_amt_data_7 * mod_data.total_rech_data_7 + mod_data.total_rech_amt_7)) / 2

mod_data.drop(['av_rech_amt_data_6','total_rech_data_6','total_rech_amt_6','av_rech_amt_data_7',
               'total_rech_data_7','total_rech_amt_7'], axis=1, inplace=True)

high_value_cust = mod_data[mod_data.av_rech_amt_6_7>mod_data.av_rech_amt_6_7.quantile(0.7)]
len(high_value_cust)

29953

In [166]:
high_value_cust.shape

(29953, 207)

**Tagging the churned customers (churn=1, else 0) based on the fourth month as follows: Those who have not made any calls (either incoming or outgoing) AND have not used mobile internet even once in the churn phase. The attributes you need to use to tag churners are:**
<br>
    1. total_ic_mou_9
    2. total_og_mou_9
    3. vol_2g_mb_9
    4. vol_3g_mb_9

In [167]:
high_value_cust['churn'] = (high_value_cust.total_ic_mou_9+high_value_cust.total_og_mou_9 + high_value_cust.vol_3g_mb_9 + high_value_cust.vol_2g_mb_9).apply(lambda x: 1 if x==0 else 0)
high_value_cust.head()

loc_og_t2o_mou  std_og_t2o_mou  loc_ic_t2o_mou    arpu_6    arpu_7  \
0              0.0             0.0             0.0   197.385   214.816   
7              0.0             0.0             0.0  1069.180  1349.850   
8              0.0             0.0             0.0   378.721   492.223   
21             0.0             0.0             0.0   514.453   597.753   
23             0.0             0.0             0.0    74.350   193.897   

      arpu_8   arpu_9  onnet_mou_6  onnet_mou_7  onnet_mou_8  onnet_mou_9  \
0    213.803   21.100        34.31        32.33         0.00        29.84   
7   3171.480  500.000        57.84        54.68        52.29        29.84   
8    137.362  166.787       413.69       351.03        35.08        33.46   
21   637.760  578.596       102.41       132.11        85.14       161.63   
23   366.966  811.480        48.96        50.66        33.58        15.74   

    offnet_mou_6  offnet_mou_7  offnet_mou_8  offnet_mou_9  roam_ic_mou_6  \
0          96.31        91.735          0.00         87.29           0.00   
7         453.43       567.160        325.91         87.29          16.23   
8          94.66        80.630        136.48        108.71           0.00   
21        757.93       896.680        983.39        869.89           0.00   
23         85.41        89.360        205.89        111.88           0.00   

    roam_ic_mou_7  roam_ic_mou_8  roam_ic_mou_9  roam_og_mou_6  roam_og_mou_7  \
0            0.00           0.00            0.0           0.00           0.00   
7           33.49          31.64            0.0          23.74          12.59   
8            0.00           0.00            0.0           0.00           0.00   
21           0.00           0.00            0.0           0.00           0.00   
23           0.00           0.00            0.0           0.00           0.00   

    roam_og_mou_8  roam_og_mou_9  loc_og_t2t_mou_6  loc_og_t2t_mou_7  \
0            0.00            0.0             11.91             11.61   
7           38.06            0.0             51.39             31.38   
8            0.00            0.0            297.13            217.59   
21           0.00            0.0              4.48              6.16   
23           0.00            0.0             48.96             50.66   

    loc_og_t2t_mou_8  loc_og_t2t_mou_9  loc_og_t2m_mou_6  loc_og_t2m_mou_7  \
0               0.00             11.26             41.03             40.43   
7              40.28             11.26            308.63            447.38   
8              12.49             26.13             80.96             70.58   
21             23.34             29.98             91.81             87.93   
23             33.58             15.74             82.94             83.01   

    loc_og_t2m_mou_8  loc_og_t2m_mou_9  loc_og_t2f_mou_6  loc_og_t2f_mou_7  \
0               0.00             39.12              0.00              0.00   
7             162.28             39.12             62.13             55.14   
8              50.54             34.58              0.00              0.00   
21            104.81            107.54              0.75              0.00   
23            148.56             83.73              0.00              0.00   

    loc_og_t2f_mou_8  loc_og_t2f_mou_9  loc_og_t2c_mou_6  loc_og_t2c_mou_7  \
0               0.00              0.00               0.0               0.0   
7              53.23              0.00               0.0               0.0   
8               0.00              0.00               0.0               0.0   
21              1.58              0.00               0.0               0.0   
23              0.00              2.45               0.0               0.0   

    loc_og_t2c_mou_8  loc_og_t2c_mou_9  loc_og_mou_6  loc_og_mou_7  \
0               0.00              0.00         65.11        63.685   
7               0.00              0.00        422.16       533.910   
8               7.15              0.00        378.09       288.180   
21              0.00      

In [168]:
high_value_cust.churn.value_counts()

0    27520
1     2433
Name: churn, dtype: int64

In [169]:
print('churn rate:', round((2433/27520)*100,2), '%')

churn rate: 8.84 %


Our dataset has high class imbalance, we will take care of it while building a model.

Removing all the attributes corresponding to the churn phase (all attributes having ‘ _9’, etc. in their names).

In [170]:
high_value_cust.drop(list(filter(lambda x: x[-1]=='9',high_value_cust.columns)), axis=1, inplace=True)
high_value_cust.head()

loc_og_t2o_mou  std_og_t2o_mou  loc_ic_t2o_mou    arpu_6    arpu_7  \
0              0.0             0.0             0.0   197.385   214.816   
7              0.0             0.0             0.0  1069.180  1349.850   
8              0.0             0.0             0.0   378.721   492.223   
21             0.0             0.0             0.0   514.453   597.753   
23             0.0             0.0             0.0    74.350   193.897   

      arpu_8  onnet_mou_6  onnet_mou_7  onnet_mou_8  offnet_mou_6  \
0    213.803        34.31        32.33         0.00         96.31   
7   3171.480        57.84        54.68        52.29        453.43   
8    137.362       413.69       351.03        35.08         94.66   
21   637.760       102.41       132.11        85.14        757.93   
23   366.966        48.96        50.66        33.58         85.41   

    offnet_mou_7  offnet_mou_8  roam_ic_mou_6  roam_ic_mou_7  roam_ic_mou_8  \
0         91.735          0.00           0.00           0.00           0.00   
7        567.160        325.91          16.23          33.49          31.64   
8         80.630        136.48           0.00           0.00           0.00   
21       896.680        983.39           0.00           0.00           0.00   
23        89.360        205.89           0.00           0.00           0.00   

    roam_og_mou_6  roam_og_mou_7  roam_og_mou_8  loc_og_t2t_mou_6  \
0            0.00           0.00           0.00             11.91   
7           23.74          12.59          38.06             51.39   
8            0.00           0.00           0.00            297.13   
21           0.00           0.00           0.00              4.48   
23           0.00           0.00           0.00             48.96   

    loc_og_t2t_mou_7  loc_og_t2t_mou_8  loc_og_t2m_mou_6  loc_og_t2m_mou_7  \
0              11.61              0.00             41.03             40.43   
7              31.38             40.28            308.63            447.38   
8             217.59             12.49             80.96             70.58   
21              6.16             23.34             91.81             87.93   
23             50.66             33.58             82.94             83.01   

    loc_og_t2m_mou_8  loc_og_t2f_mou_6  loc_og_t2f_mou_7  loc_og_t2f_mou_8  \
0               0.00              0.00              0.00              0.00   
7             162.28             62.13             55.14             53.23   
8              50.54              0.00              0.00              0.00   
21            104.81              0.75              0.00              1.58   
23            148.56              0.00              0.00              0.00   

    loc_og_t2c_mou_6  loc_og_t2c_mou_7  loc_og_t2c_mou_8  loc_og_mou_6  \
0                0.0               0.0              0.00         65.11   
7                0.0               0.0              0.00        422.16   
8                0.0               0.0              7.15        378.09   
21               0.0               0.0              0.00         97.04   
23               0.0               0.0             17.71        131.91   

    loc_og_mou_7  loc_og_mou_8  std_og_t2t_mou_6  std_og_t2t_mou_7  \
0         63.685          0.00              0.00              0.00   
7        533.910        255.79              4.30             23.29   
8        288.180         63.04            116.56            133.43   
21        94.090        129.74             97.93            125.94   
23       133.680        182.14              0.00              0.00   

    std_og_t2t_mou_8  std_og_t2m_mou_6  std_og_t2m_mou_7  std_og_t2m_mou_8  \
0               0.00              3.95             3.635              0.00   
7              12.01             49.89            31.760             49.14   
8              22.58             13.69            10.040             75.69   
21             61.79            665.36           808.740            876.99   
23              0.00              2.36             6.350    

In [171]:
high_value_cust.shape

(29953, 157)

As checked in the data dictionary, columns start with fb and night are schemes which are used for facebook and night packs respectively, so they are categorical columns(yes/no). Same as with churn columns. We will convert then to object type. This will help in doing EDA. 

In [172]:
cols=list(filter(lambda x: x.startswith('fb') or x.startswith('night'), high_value_cust.columns))
cols

['night_pck_user_6',
 'night_pck_user_7',
 'night_pck_user_8',
 'fb_user_6',
 'fb_user_7',
 'fb_user_8']

In [173]:
cols.append('churn')
high_value_cust[cols]=high_value_cust[cols].astype('object')

### Exploratory Data Analysis

Let us create some utility functions

In [174]:
# Method to add or subtract 2 columns to form a new column. It also have a feature to add or subtract 2 columns 
# based on complete column name or a pattern provided.
    # col_a_end_str - column name or pattern from end for column A
    # col_b_end_str - column name or pattern from end for column B
    # avg_or_diff - 'avg' for average and 'diff' for subtraction of 2 columns
    # new_name_end_str - end pattern to give to new column/s
    # dataframe - a dataframe
    # complete_column_name_given - flag to check if willing to merge all the columns based on pattern or not    

def addOrSubColumns(col_a_end_str, col_b_end_str, avg_or_diff, new_name_end_str, dataframe, complete_column_name_given=False):
    li=[]
    if complete_column_name_given:
        new_name= col_a_end_str+'_'+col_b_end_str+'_'+new_name_end_str
        if avg_or_diff=='diff':
                dataframe[new_name]= (dataframe[col_b_end_str] - dataframe[col_a_end_str])
        else:
            dataframe[new_name]= (dataframe[col_b_end_str] + dataframe[col_a_end_str])/2
        print(new_name)
        li+=[col_a_end_str,col_b_end_str]

    else:
        s=set(filter( lambda x: x[-len(col_a_end_str):]==col_a_end_str, dataframe.columns))
        s1=set(filter( lambda x:  x[-len(col_b_end_str):]==col_b_end_str, dataframe.columns))
        
        for i in list(s):
            k=i[:-len(col_a_end_str)]
            a=k+col_a_end_str
            b=k+col_b_end_str
            if  b in s1:
                if avg_or_diff=='diff':
                    dataframe[k+new_name_end_str]= (dataframe[b] - dataframe[a])
                else:
                    dataframe[k+new_name_end_str]= (dataframe[b] + dataframe[a])/2
                li+=[a,b]
                s.remove(a); s1.remove(b)
        
    return dataframe.drop(li, axis=1)

In [175]:
# ---- Univariate Analysis ---- #
def univariate(dataset):
    col = dataset.columns
    for i in col:
        plt.figure(figsize=(12, 6))
        if dataset[i].dtypes != 'object':
            sns.distplot(i)
            i.describe()
        else:
            sns.countplot(i)
            i.value_counts()

In [176]:
# ---- Bivariate Analysis ---- #
def bivariate(dataset,col1, col2):
    if dataset[col1].dtypes != 'object' and dataset[col2].dtypes != 'object':
        sns.regplot(col1, col2)
    elif (dataset[col1].dtypes == 'object' and dataset[col2].dtypes != 'object') or (dataset[col1].dtypes != 'object' and dataset[col2].dtypes == 'object'):
        g = sns.boxplot(col1, col2)
#         if dataset[col2].nunique()>10:
#                 g.set_xticklabels(g.get_xticklabels(),rotation=90)

In [177]:
# ---- Bivariate Analysis with Churn as one of the column ---- #
def bivariate_churn(dataset,col):
    if dataset[col].dtypes != 'object':
        sns.boxplot(col, 'churn')

In [205]:
# method to cap outliers
def capingOutliers(dataframe, quantile, columns, cap=False):
    for i in columns:
        print('outliers in',i, ':', len(dataframe[i][dataframe[i]>dataframe[i].quantile(quantile)]))
        if cap:
            dataframe[i][dataframe[i]>dataframe[i].quantile(quantile)] = dataframe[i].quantile(quantile)

In [178]:
# -- do some analysis

__Outlier Treatment__

In [212]:
round(high_value_cust.describe(percentiles=[0.25,0.5,0.75,0.9,0.95,0.99]),2)

aon  std_ic_t2f_mou_diff  total_rech_num_diff  arpu_2g_diff  \
count  29953.00             29953.00             29953.00      29953.00   
mean    1193.24                -0.98                -2.10        -13.09   
std      920.32                 8.92                 5.96         91.47   
min      180.00              -636.50              -147.50      -2980.04   
25%      460.00                -0.12                -4.00        -10.60   
50%      846.00                 0.00                -1.50          0.00   
75%     1756.00                 0.00                 1.00          0.00   
90%     2793.00                 1.25                 4.50         56.87   
95%     3140.00                 4.80                 7.50        142.79   
99%     3140.00                 4.80                 7.50        142.81   
max     3140.00                 4.80                 7.50        142.81   

       std_ic_t2t_mou_diff  monthly_2g_diff  roam_ic_mou_diff  \
count             29953.00         29953.00          29953.00   
mean                 -4.44            -0.05             -6.91   
std                  42.46             0.29             44.63   
min               -2897.68            -3.00          -1296.99   
25%                  -2.83             0.00              0.00   
50%                   0.00             0.00              0.00   
75%                   0.33             0.00              0.00   
90%                   8.95             0.00              4.02   
95%                  21.71             0.50             26.24   
99%                  21.71             0.50             26.24   
max                  21.71             0.50             26.24   

       monthly_3g_diff  std_ic_t2m_mou_diff  count_rech_2g_diff  \
count         29953.00             29953.00            29953.00   
mean             -0.03                -6.48               -0.22   
std               0.40                52.69                1.54   
min              -7.50             -2624.90              -25.50   
25%               0.00                -7.36               -0.50   
50%               0.00                -0.24                0.00   
75%               0.00                 2.78                0.00   
90%               0.00                19.55                1.00   
95%               1.00                41.00                2.50   
99%               1.00                41.00                2.50   
max               1.00                41.00                2.50   

       offnet_mou_diff  vol_3g_mb_diff  total_og_mou_diff  og_others_diff  \
count         29953.00        29953.00           29953.00        29953.00   
mean            -58.71          -48.78             -95.32           -0.33   
std             278.63          465.25             419.34            2.09   
min           -7094.39       -12385.50           -7213.41         -304.90   
25%             -98.88           -2.40            -167.86            0.00   
50%             -10.93            0.00             -14.62            0.00   
75%              44.38            0.00              67.92            0.00   
90%             171.03          312.17             291.39            0.00   
95%             328.75          702.16             539.39            0.00   
99%             328.78          702.25             539.46            0.00   
max             328.78          702.25             539.46            0.00   

       loc_og_t2c_mou_diff  loc_ic_mou_diff  count_rech_3g_diff  \
count             29953.00         29953.00            29953.00   
mean                 -0.51           -18.07               -0.09   
std                   5.11           144.23                0.70   
min                -384.39         -3747.23              -21.50   
25%                  -0.36           -48.46                0.00   
50%                   0.00            -3.27                0.00   
75%                   0.00            38.34                0.00   
90%                   2.60           111.64                0.50 

In [ ]:
# box plots to analyse the outliers

In [211]:
# call the function to check and cap the outliers
cols=list(high_value_cust.select_dtypes(exclude='object').columns) # columns to remove ouliers
capingOutliers(high_value_cust, 0.95, cols, True)

outliers in aon : 1490
outliers in std_ic_t2f_mou_diff : 0
outliers in total_rech_num_diff : 0
outliers in arpu_2g_diff : 1498
outliers in std_ic_t2t_mou_diff : 1498
outliers in monthly_2g_diff : 1115
outliers in roam_ic_mou_diff : 1497
outliers in monthly_3g_diff : 447
outliers in std_ic_t2m_mou_diff : 1498
outliers in count_rech_2g_diff : 1365
outliers in offnet_mou_diff : 1498
outliers in vol_3g_mb_diff : 1498
outliers in total_og_mou_diff : 1498
outliers in og_others_diff : 127
outliers in loc_og_t2c_mou_diff : 1496
outliers in loc_ic_mou_diff : 1498
outliers in count_rech_3g_diff : 972
outliers in spl_og_mou_diff : 1498
outliers in onnet_mou_diff : 1498
outliers in isd_ic_mou_diff : 1498
outliers in loc_ic_t2f_mou_diff : 1498
outliers in loc_og_t2f_mou_diff : 1498
outliers in std_og_t2t_mou_diff : 1498
outliers in sachet_3g_diff : 1359
outliers in ic_others_diff : 1498
outliers in std_og_t2f_mou_diff : 1495
outliers in sachet_2g_diff : 1381
outliers in loc_og_t2m_mou_diff : 1498
o

In [179]:
# -- do some analysis

In [180]:
# aggregating the columns of good months
high_value_cust=addOrSubColumns('6','7','avg','6_7',high_value_cust)

In [181]:
# -- do some analysis

In [182]:
# getting average recharge amount for action month
high_value_cust['av_rech_amt_8']=(high_value_cust.av_rech_amt_data_8 * high_value_cust.total_rech_data_8 + 
                                  high_value_cust.total_rech_amt_8)

high_value_cust.drop(['av_rech_amt_data_8','total_rech_data_8','total_rech_amt_8'], axis=1, inplace=True)

In [183]:
# -- do some analysis

In [184]:
# difference of the columns between action month and average of good month
high_value_cust=addOrSubColumns('6_7','8','diff','diff',high_value_cust)

In [185]:
# -- do some analysis

In [186]:
# -- do some analysis

Removing the columns having more than 85% of values as a single value (highly skewed columns)

In [187]:
li=[]
for i in high_value_cust.columns:
    if max(high_value_cust[i].value_counts())/len(high_value_cust) >0.85:
        li.append(i)

li.remove('churn')
high_value_cust.drop(li, axis=1, inplace=True)

In [188]:
high_value_cust.shape

(29953, 48)

## Model Building

In [ ]:
# import required libraries
from sklearn.model_selection import train_test_split
from sklearn.pipeline import Pipeline
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler, MinMaxScaler
# from sklearn.pipeline import FeatureUnion
# from sklearn.base import BaseEstimator, TransformerMixin
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report
from sklearn.metrics import roc_auc_score
# from sklearn.metrics import sensitivity_specificity_support
from sklearn.model_selection import StratifiedKFold
from sklearn.model_selection import cross_val_score
from sklearn.metrics import confusion_matrix
from sklearn.model_selection import GridSearchCV
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.svm import SVC

In [ ]:
high_value_cust.columns

In [ ]:
# divide data into train and test
X = high_value_cust.drop("churn", axis = 1)
y = high_value_cust.churn
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25, random_state = 42)

In [ ]:
# print shapes of train and test sets
print(X_train.shape)
print(y_train.shape)
print(X_test.shape)
print(y_test.shape)

In [ ]:
# apply pca to train data
pca = Pipeline([('scaler', MinMaxScaler()), ('pca', PCA())])

In [ ]:
pca.fit(X_train)
churn_pca = pca.fit_transform(X_train)

In [ ]:
# extract pca model from pipeline
pca = pca.named_steps['pca']

# look at explainded variance of PCA components
print(pd.Series(np.round(pca.explained_variance_ratio_.cumsum(), 4)*100))

In [ ]:
# plot feature variance
features = range(pca.n_components_)
cumulative_variance = np.round(np.cumsum(pca.explained_variance_ratio_)*100, decimals=4)
plt.figure(figsize=(175/20,100/20)) # 100 elements on y-axis; 175 elements on x-axis; 20 is normalising factor
plt.plot(cumulative_variance)

In [ ]:
# create pipeline
PCA_VARS = 20
steps = [('scaler', MinMaxScaler()),
         ("pca", PCA(n_components=PCA_VARS)),
         ("logistic", LogisticRegression(class_weight='balanced'))
        ]
pipeline = Pipeline(steps)

In [ ]:
# fit model
pipeline.fit(X_train, y_train)

# check score on train data
pipeline.score(X_train, y_train)

In [ ]:
# predict churn on test data
y_pred = pipeline.predict(X_test)

# create onfusion matrix
cm = confusion_matrix(y_test, y_pred)
print(cm)

# # check sensitivity and specificity
# sensitivity, specificity, _ = sensitivity_specificity_support(y_test, y_pred, average='binary')
# print("Sensitivity: \t", round(sensitivity, 2), "\n", "Specificity: \t", round(specificity, 2), sep='')

# check area under curve
y_pred_prob = pipeline.predict_proba(X_test)[:, 1]
print("AUC:    \t", round(roc_auc_score(y_test, y_pred_prob),2))

In [ ]:
# PCA
pca = PCA()

# logistic regression - the class weight is used to handle class imbalance - it adjusts the cost function
logistic = LogisticRegression(class_weight={0:0.1, 1: 0.9})

# create pipeline
steps = [("scaler", MinMaxScaler()), 
         ("pca", pca),
         ("logistic", logistic)
        ]

# compile pipeline
pca_logistic = Pipeline(steps)

# hyperparameter space
params = {'pca__n_components': [20, 30], 'logistic__C': [0.1, 0.5, 1, 2, 3, 4, 5, 10], 'logistic__penalty': ['l1', 'l2']}

# create 5 folds
folds = StratifiedKFold(n_splits = 5, shuffle = True, random_state = 4)

# create gridsearch object
model = GridSearchCV(estimator=pca_logistic, cv=folds, param_grid=params, scoring='roc_auc', n_jobs=-1, verbose=1)

In [ ]:
# fit model
model.fit(X_train, y_train)

In [ ]:
# cross validation results
pd.DataFrame(model.cv_results_)

In [ ]:
# print best hyperparameters
print("Best AUC: ", model.best_score_)
print("Best hyperparameters: ", model.best_params_)

In [ ]:
# predict churn on test data
y_pred = model.predict(X_test)

# create onfusion matrix
cm = confusion_matrix(y_test, y_pred)
print(cm)

# # check sensitivity and specificity
# sensitivity, specificity, _ = sensitivity_specificity_support(y_test, y_pred, average='binary')
# print("Sensitivity: \t", round(sensitivity, 2), "\n", "Specificity: \t", round(specificity, 2), sep='')

# check area under curve
y_pred_prob = model.predict_proba(X_test)[:, 1]
print("AUC:    \t", round(roc_auc_score(y_test, y_pred_prob),2))

In [ ]:
# random forest - the class weight is used to handle class imbalance - it adjusts the cost function
forest = RandomForestClassifier(class_weight={0:0.1, 1: 0.9}, n_jobs = -1)

# hyperparameter space
params = {"criterion": ['gini', 'entropy'], "max_features": ['auto', 0.4]}

# create 5 folds
folds = StratifiedKFold(n_splits = 5, shuffle = True, random_state = 4)

# create gridsearch object
model = GridSearchCV(estimator=forest, cv=folds, param_grid=params, scoring='roc_auc', n_jobs=-1, verbose=1)

In [ ]:
# fit model
model.fit(X_train, y_train)

In [ ]:
# print best hyperparameters
print("Best AUC: ", model.best_score_)
print("Best hyperparameters: ", model.best_params_)

In [ ]:
# predict churn on test data
y_pred = model.predict(X_test)

# create onfusion matrix
cm = confusion_matrix(y_test, y_pred)
print(cm)

# # check sensitivity and specificity
# sensitivity, specificity, _ = sensitivity_specificity_support(y_test, y_pred, average='binary')
# print("Sensitivity: \t", round(sensitivity, 2), "\n", "Specificity: \t", round(specificity, 2), sep='')

# check area under curve
y_pred_prob = model.predict_proba(X_test)[:, 1]
print("AUC:    \t", round(roc_auc_score(y_test, y_pred_prob),2))

In [ ]:
# run a random forest model on train data
max_features = int(round(np.sqrt(X_train.shape[1])))    # number of variables to consider to split each node
print(max_features)

rf_model = RandomForestClassifier(n_estimators=100, max_features=max_features, class_weight={0:0.1, 1: 0.9}, oob_score=True, random_state=4, verbose=1)

In [ ]:
# fit model
rf_model.fit(X_train, y_train)

In [ ]:
# OOB score
rf_model.oob_score_

In [ ]:
# predict churn on test data
y_pred = rf_model.predict(X_test)

# create onfusion matrix
cm = confusion_matrix(y_test, y_pred)
print(cm)

# # check sensitivity and specificity
# sensitivity, specificity, _ = sensitivity_specificity_support(y_test, y_pred, average='binary')
# print("Sensitivity: \t", round(sensitivity, 2), "\n", "Specificity: \t", round(specificity, 2), sep='')

# check area under curve
y_pred_prob = rf_model.predict_proba(X_test)[:, 1]
print("ROC:    \t", round(roc_auc_score(y_test, y_pred_prob),2))

In [ ]:
# predictors
features = high_value_cust.drop('churn', axis=1).columns

# feature_importance
importance = rf_model.feature_importances_

# create dataframe
feature_importance = pd.DataFrame({'variables': features, 'importance_percentage': importance*100})
feature_importance = feature_importance[['variables', 'importance_percentage']]

# sort features
feature_importance = feature_importance.sort_values('importance_percentage', ascending=False).reset_index(drop=True)
print("Sum of importance=", feature_importance.importance_percentage.sum())
feature_importance

In [ ]:
# extract top 'n' features
top_n = 10
top_features = feature_importance.variables[0:top_n]

In [ ]:
# plot feature correlation
import seaborn as sns
plt.rcParams["figure.figsize"] =(15,10)
# mycmap = sns.diverging_palette(199, 359, s=99, center="light", as_cmap=True)
sns.heatmap(data=X_train[top_features].corr(), center=0.0, annot=True)

In [ ]:
X_train = X_train[top_features]
X_test = X_test[top_features]

In [ ]:
# logistic regression
steps = [('scaler', MinMaxScaler()), 
         ("logistic", LogisticRegression(class_weight={0:0.1, 1:0.9}))
        ]

# compile pipeline
logistic = Pipeline(steps)

# hyperparameter space
params = {'logistic__C': [0.1, 0.5, 1, 2, 3, 4, 5, 10], 'logistic__penalty': ['l1', 'l2']}

# create 5 folds
folds = StratifiedKFold(n_splits = 5, shuffle = True, random_state = 4)

# create gridsearch object
model = GridSearchCV(estimator=logistic, cv=folds, param_grid=params, scoring='roc_auc', n_jobs=-1, verbose=1)

In [ ]:
# fit model
model.fit(X_train, y_train)

In [ ]:
# print best hyperparameters
print("Best AUC: ", model.best_score_)
print("Best hyperparameters: ", model.best_params_)

In [ ]:
# predict churn on test data
y_pred = model.predict(X_test)

# create onfusion matrix
cm = confusion_matrix(y_test, y_pred)
print(cm)

# # check sensitivity and specificity
# sensitivity, specificity, _ = sensitivity_specificity_support(y_test, y_pred, average='binary')
# print("Sensitivity: \t", round(sensitivity, 2), "\n", "Specificity: \t", round(specificity, 2), sep='')

# check area under curve
y_pred_prob = model.predict_proba(X_test)[:, 1]
print("ROC:    \t", round(roc_auc_score(y_test, y_pred_prob),2))